src/video.rs

```diff
@@ -115,34 +115,61 @@ fn pretty_spinner_style() -> ProgressStyle {
         .progress_chars(PROGRESS_CHARS)
 }
 
+type VideoCompareMutex<E, F> = Arc<Mutex<VideoCompare<E, F>>>;
+
+struct VideoCompare<E: Decoder, F: Decoder> {
+    current_frame: usize,
+    next_frame: usize,
+    source: E,
+    distorted: F,
+}
+
 fn calc_score<S: Pixel, D: Pixel, E: Decoder, F: Decoder>(
-    mtx: &Mutex<(usize, (E, F))>,
+    mtx: &VideoCompareMutex<E, F>,
     src_yuvcfg: &YuvConfig,
     dst_yuvcfg: &YuvConfig,
     inc: usize,
+    end_frame: Option<usize>,
     verbose: bool,
 ) -> Option<(usize, f64)> {
     let (frame_idx, (src_frame, dst_frame)) = {
         let mut guard = mtx.lock().unwrap();
-        let curr_frame = guard.0;
+        let mut curr_frame = guard.current_frame;
+        // We passed this as start frame.
+        // However, we are going to use it to store the next frame we should compute.
+        let mut next_frame = guard.next_frame;
+
+        //println!("{curr_frame} < {next_frame}");
+
+        while curr_frame < next_frame {
```

Comment:
I don't like that we need to have mutable indices here. `calc_score` is becoming a bit complex. Can we do something like

```rs
let skip_frames_until_next = guard.next_frame - guard.current_frame;
for ii in 0..skip_frames_until_next {
  // skip frames
}

let src_frame = guard.source.read_video_frame::<S>();
let dst_frame = guard.source.read_video_frame::<D>();

guard.current_frame = 1 + guard.next_frame;
guard.next_frame += inc;
```

Should be logically equivalent and a bit easier to parse.

Reply:
I changed the logic to basically do this now.